Refactor helm charts

wilmardo opened the pull request:
# Describe what this PR does #

* Make the helm chart compatible with both Kubernetes versions.
* Bonus is that the generated templates by Helm are now linted by yamllint :)
* All the ports to variables (http-metrics and grpc metrics), chose to provide a service for each since Prometheus annotations do not support multiple ports (see [prometheus/prometheus#3756](https://github.com/prometheus/prometheus/issues/3756))

Fixes: #565
Implements changes discussed in: #555 

* ~Generate deployment files to keep the maintaince of changes in one place (the helm chart) to avoid errors and changed not being implement in one of the versions as we saw with #555~  
* ~Generated deployments deploy to the ceph-csi namespace which is created on deployment.~
Both split to #604

Madhu-1 commented:
yes kube 1.13 need lesser permission compare to 1.4 and 1.15

wilmardo commented:
@Madhu-1 @humblec (others might chime in 

What do you guys think of the current approach? Before I refactor the e2e tests and the readme :)
If this is something to move forward with, what are sensible values.yaml files to generate the deployments with?

wilmardo commented:
@Madhu-1 @humblec You guys had any time to look at this?

I think this PR should have more importance since PR's like #575 and #566 could benefit immensely from this change. No more editing multiple files by (which is super error prone)

humblec commented:
@wilmardo the PR is `WIP` and tests are failing too. Can you please check on these?

wilmardo commented:
@humblec Please read my comment above, I did not refactor the tests just yet since this is such a huge change in workflow and repository setup that I first want some opinions before I move on with this.

Madhu-1 commented:
@wilmardo will review this PR soon, Thanks for the  PR :+1:

Madhu-1 commented:
@wilmardo can you add a readme or PR description how this works, that would really help a lot to understand and review this.

wilmardo commented:
@Madhu-1 Added a the **How to test this PR** section to the OP.
If any questions arise, please post a comment. Github is my daily hangout ;)

Madhu-1 commented:
@wilmardo JFYI we have cephcsi  slack channel where most dev's are available

Madhu-1 commented:
@wilmardo moving this out of release 1.2.0 as discussed in  slack

Madhu-1 commented on scripts/gen-deploy.sh:
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
```
change file permission, not  able to execute it

wilmardo replied:
Done, Windows devstation things :')

Madhu-1 commented:
@wilmardo do we need to make changes to the kube templates here? with this changes all kube templates will be merged into one file and based on the values you are generating 4 different files for each kube version

```
without-rbac-without-attacher.yaml
without-rbac.yaml
with-rbac-without-attacher.yaml
with-rbac.yaml
```
my suggestion will be to make changes only to the helm chats and leave kube templates as it is.

@humblec @ShyamsundarR any thoughts?

humblec commented:
> @wilmardo do we need to make changes to the kube templates here? with this changes all kube templates will be merged into one file and based on the values you are generating 4 different files for each kube version
> 
> ```
> without-rbac-without-attacher.yaml
> without-rbac.yaml
> with-rbac-without-attacher.yaml
> with-rbac.yaml
> ```
> 
> my suggestion will be to make changes only to the helm chats and leave kube templates as it is.
> 
> @humblec @ShyamsundarR any thoughts?

Yeah, agree, lets not change the kube templates now as we do have references and those are used in rook..etc

wilmardo commented:
Hmm, that still leaves the problem of maintaining 8 files for one change (which results in human errors as seen before). Files to edit after a change to the provisioner:

* cephfs/1.13/provisioner-statefulset
* cephfs/1.14/provisioner-deployment
* rbd/1.13/provisioner-statefulset
* rbd/1.14/provisioner-deployment
* charts/ceph-csi-cephfs/templates/provisioner-deployment
* charts/ceph-csi-cephfs/templates/provisioner-statefulset
* charts/ceph-csi-rbd/templates/provisioner-deployment
* charts/ceph-csi-rbd/templates/provisioner-statefulset

That is the main reason for this whole refactor, I could split it to a seperate PR? But I strongly suggest that it has priority. The generation of deployment files could half these changes, the only duplication is cephfs/rbd but that is inevitable.

Madhu-1 commented:
yes can you split it to separate PR for kube template update and 
i would like to see two files in kubernetes folder with attacher and without attacher in separate PR to generate kube templates from helm chats

Madhu-1 commented on charts/ceph-csi-cephfs/templates/csidriver-crd.yaml:
```diff
@@ -8,4 +7,4 @@ metadata:
   attachRequired: false
   podInfoOnMount: false
-{{ end }}
+{{- end -}}
```
GitHub is showing missing newline at end of file

wilmardo replied:
I removed all newlines since they are useless in the templates, the -}} would chomp it anyways

Madhu-1 replied:
this is required to make Github happy, can we revert it back?

wilmardo replied:
Will do :)

Madhu-1 commented on charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml:
```diff
@@ -59,24 +59,23 @@ spec:
+          args:
             - "--nodeid=$(NODE_ID)"
             - "--type=cephfs"
+{{- if or (eq .Values.nodeplugin.plugin.image.tag "canary") (semverCompare ">=1.2.0" .Values.nodeplugin.plugin.image.tag | default 0) }}
```
remove if the check for this one as we release helm templates for each version

wilmardo replied:
Wouldn't it be nice to support v1.1.0 images with this new charts?

Madhu-1 replied:
it recommended using helm chats from the released version. not to use 1.2.0 helm chats for 1.1.0 deployments

wilmardo replied:
Nah I get the point, it would introduce a spaghetti of if statements, will remove :)

Madhu-1 commented on charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml:
```diff
@@ -117,16 +120,21 @@ spec:
             - "--type=liveness"
             - "--endpoint=$(CSI_ENDPOINT)"
-            - "--metricsport=8081"
+{{- if or (eq .Values.nodeplugin.plugin.image.tag "canary") (semverCompare ">=1.2.0" .Values.nodeplugin.plugin.image.tag | default 0) }}
```
same here and controllerser and pidlimit does not apply for liveness container

Madhu-1 commented on charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml:
```diff
@@ -59,24 +59,23 @@ spec:
+{{- if or (eq .Values.nodeplugin.plugin.image.tag "canary") (semverCompare ">=1.2.0" .Values.nodeplugin.plugin.image.tag | default 0) }}
             - "--nodeserver=true"
+            - "--pidlimit=-1"
+            - "--metricsport={{ .Values.nodeplugin.grpcMetrics.containerPort }}"
```
if grpc metrics is enabled then set this flags

Madhu-1 commented on charts/ceph-csi-cephfs/templates/nodeplugin-grpc-service.yaml:
```diff
@@ -0,0 +1,41 @@
+    heritage: {{ .Release.Service }}
+spec:
+{{- if .Values.nodeplugin.grpcMetrics.service.clusterIP }}
+  clusterIP: "{{ .Values.nodeplugin.grpcMetrics.service.clusterIP }}"
```
this need to validate
cannot use cluster IP for node plugin I believe as this is because host networking is enabled

wilmardo replied:
I am testing this as we speak

wilmardo replied:
Just deployed the chart with:

```yaml
---
csiConfig:
  - clusterID: <redacted>
    monitors:
      - 192.168.2.100:6789

nodeplugin:
  httpMetrics:
    enabled: true
    service:
      type: ClusterIP

  grpcMetrics:
    enabled: true
    service:
      type: ClusterIP

provisioner:
  httpMetrics:
    enabled: true
    service:
      type: ClusterIP

  grpcMetrics:
    enabled: true
    service:
      type: ClusterIP
```

And I can curl cluster-ip:8090/metrics and cluster-ip:8080/metrics perfectly from within the cluster for both the nodeplugin and the provisioner :)

Madhu-1 replied:
Thanks for testing it

wilmardo commented:
@Madhu-1 Split the deployment changes into #604 and rebased to current master

Madhu-1 commented:
> @Madhu-1 Split the deployment changes into #604 and rebased to current master

Awesome  :+1:

Madhu-1 commented:
@wilmardo you need to update https://github.com/ceph/ceph-csi/blob/master/deploy.sh#L3 function also

Madhu-1 commented:
@wilmardo lets push canary versioned helm chats from master branch this is intended to updated on each build from master and will change it to release version tag in the release
https://github.com/ceph/ceph-csi/blob/49caa8ea01472c08080deaa1bad3e3dfcc1c5ad3/charts/ceph-csi-cephfs/Chart.yaml#L7

Madhu-1 commented:
need to change sources https://github.com/ceph/ceph-csi/blob/master/deploy/cephfs/kubernetes/v1.13/helm/ceph-csi-cephfs/Chart.yaml#L14

Madhu-1 commented:
@wilmardo let us know when it is ready for review

wilmardo commented:
@Madhu-1 Implemented the changes. Each master build releases the Chart.yaml as available on the master:
https://github.com/wilmardo/ceph-csi/blob/improve-helm/charts/ceph-csi-cephfs/Chart.yaml

On each tagged run of deploy.sh the version, appVersion and sources are updated with sed.
For example a run for the v1.2.1 tag should look like this:

```yaml
---
apiVersion: v1
appVersion: 1.2.1
description: "Container Storage Interface (CSI) driver,
provisioner, and attacher for Ceph cephfs"
name: ceph-csi-cephfs
version: 1.2.1
keywords:
  - ceph
  - cephfs
  - ceph-csi
home: https://github.com/ceph/ceph-csi
sources:
  - https://github.com/ceph/ceph-csi/tree/release-v1.2.1/charts/ceph-csi-rbd
icon: https://raw.githubusercontent.com/ceph/ceph-csi/master/assets/ceph-logo.png

```

Madhu-1 commented on charts/ceph-csi-cephfs/Chart.yaml:
```diff
@@ -1,15 +1,15 @@
 provisioner, and attacher for Ceph cephfs"
 name: ceph-csi-cephfs
-version: 0.9.0
+version: 1.3.0-canary
```
instead of 1.3.0 can this be canary?

wilmardo replied:
No, helm enforces semver on this field and a string isn't valid semver

wilmardo replied:
So before a minor version release these values need to be updated. Sadly I can't make much more of this

Madhu-1 commented on charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml:
```diff
@@ -33,24 +33,24 @@ spec:
-                    'rm -rf /registration/{{ .Values.driverName }}
+                command: [
+                  "/bin/sh", "-c",
+                  'rm -rf /registration/{{ .Values.driverName }}
```
please check a space is present after driver name
this should be something like `rm  -rf /abc/a bcd/d`

wilmardo replied:
Done by removing all spaces from the string

Madhu-1 commented on charts/ceph-csi-cephfs/templates/provisioner-service.yaml:
```diff
@@ -1,3 +1,4 @@
+{{- if semverCompare "<=1.13" .Capabilities.KubeVersion.GitVersion -}}
```
even if the HTTP metrics and grpc metrics are not enabled we are creating a service for it here

wilmardo replied:
See #81

Madhu-1 replied:
AFAIK we don't need the service to be created, in the stateful set you need to mention a dummy service name.

can you check it

Madhu-1 commented on charts/ceph-csi-cephfs/templates/provisioner-statefulset.yaml:
```diff
@@ -104,16 +104,17 @@ spec:
             - "--endpoint=$(CSI_ENDPOINT)"
-            - "--metricsport=8081"
-            - "--metricspath=/metrics"
+            - "--livenessport={{ .Values.provisioner.httpMetrics.containerPort }}"
```
livenessport and livenesspath replaced by mertricsport and metricspath

wilmardo replied:
Good catch, such a big refactor and at some point you become blind to your own mistakes ;)

Madhu-1 commented on charts/ceph-csi-cephfs/templates/provisioner-statefulset.yaml:
```diff
@@ -61,30 +63,23 @@ spec:
-            - "--enablegrpcmetrics=false"
+            - "--containerized=true"
           env:
-            - name: POD_IP
```
is this removed?

wilmardo replied:
Nope, github is weirding out on the diff here:
https://github.com/wilmardo/ceph-csi/blob/improve-helm/charts/ceph-csi-cephfs/templates/provisioner-statefulset.yaml#L121

Madhu-1 commented on charts/ceph-csi-rbd/Chart.yaml:
```diff
@@ -1,15 +1,15 @@
 provisioner, snapshotter, and attacher for Ceph RBD"
 name: ceph-csi-rbd
-version: 0.9.0
+version: 1.3.0-canary
```
replace  with canary

Madhu-1 commented on charts/ceph-csi-rbd/templates/nodeplugin-daemonset.yaml:
```diff
@@ -35,24 +35,24 @@ spec:
-                    'rm -rf /registration/{{ .Values.driverName }}
+                command: [
+                  "/bin/sh", "-c",
+                  'rm -rf /registration/{{ .Values.driverName }}
```
check space is present after driver name

Madhu-1 commented on charts/ceph-csi-rbd/templates/provisioner-service.yaml:
```diff
@@ -1,3 +1,4 @@
+{{- if semverCompare "<=1.13" .Capabilities.KubeVersion.GitVersion -}}
```
need to check both HTTP and grpc are enabled or not, if not just create a dummy service?

wilmardo replied:
This is needed, a StatefulSet isn't valid without a service. See this discussion from earlier:
https://github.com/ceph/ceph-csi/issues/81

Madhu-1 commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -0,0 +1,215 @@
+  plugin:
+    image:
+      repository: quay.io/cephcsi/cephcsi
+      tag: v1.2.0
```
canary tag

Madhu-1 commented on deploy.sh:
```diff
@@ -3,15 +3,26 @@
 	PACKAGE=$1
 	CHANGED=0
-	VERSION=$(grep 'version:' deploy/"$PACKAGE"/helm/Chart.yaml | awk '{print $2}')
+	VERSION=${TRAVIS_TAG//v}  # Set version (without v prefix)
```
am not aware how TRAVIS_TAG works, can you just make use of `ENV_CSI_IMAGE_VERSION` in deploy.sh?

wilmardo replied:
The `TRAVIS_TAG` is set to the git tag else it is empty:
https://docs.travis-ci.com/user/environment-variables/#default-environment-variables

Will look into this tho, it seems that the scripts exists when it is a tagged build. TRAVIS_BRANCH is then set to TRAVIS_TAG value and will en dup in the else and exit. Is this expected?

Madhu-1 replied:
we update the csi container image (canary) once each PR is merged. in the same way we need to update the helm chat for the master branch. 

will this work like it?

I suggest let's use `ENV_CSI_IMAGE_VERSION` in this PR, will move to use TRAVIS_TAG later when we integrate Travis with CD

wilmardo commented:
@Madhu-1 Found one little improvement, the standard helm labels have changed:
https://helm.sh/docs/chart_best_practices/#standard-labels

Should I incorporate this or create an issue? Seems fitting to do this now, does anything of Rook depend on this?

Madhu-1 commented:
> @Madhu-1 Found one little improvement, the standard helm labels have changed:
> https://helm.sh/docs/chart_best_practices/#standard-labels
> 
> Should I incorporate this or create an issue? Seems fitting to do this now, does anything of Rook depend on this?

let's do it as separate PR, please go ahead and create an issue

Madhu-1 commented on charts/ceph-csi-cephfs/templates/nodeplugin-daemonset.yaml:
```diff
@@ -33,24 +33,24 @@ spec:
-                    'rm -rf /registration/{{ .Values.driverName }}
+                command: [
+                  "/bin/sh", "-c",
+                  'rm -rf /registration/{{ .Values.driverName | replace " " "" }}
```
@wilmardo can you explain how this works?

wilmardo replied:
Helm has sprig available for templating functions:
https://github.com/helm/helm/blob/master/docs/chart_template_guide/functions_and_pipelines.md

More about sprig string functions:
http://masterminds.github.io/sprig/strings.html

The replace function replaces a space `" "` with no space `""`

Will add a NOTE above it to explain :)

Madhu-1 commented on docs/deploy-cephfs.md:
````diff
@@ -190,3 +190,20 @@ Dynamically povisioned volumes are deleted by the driver, when requested to
+**Deploy Helm Chart:**
+
+```bash
+helm install charts/ceph-csi-cepfs
````
instead of this can we just point user for rbd helm readme as it contains more information

wilmardo replied:
Sure!

wilmardo commented:
@Madhu-1 Switched to ENV_CSI_IMAGE_VERSION for now, will look into the refactor of deploy.sh from issue #610

Madhu-1 commented on charts/ceph-csi-rbd/templates/csidriver-crd.yaml:
```diff
@@ -1,5 +1,4 @@
----
```
this is cephfs file not rbd file.. same for other places also when I deploy rbd from helm chat am seeing cephfs templates inside it please check.

wilmardo replied:
This was due the values.yaml containing wrong values, has been fixed in 7761328f5992b213caf73c0c710cb15481767485

Madhu-1 commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -0,0 +1,215 @@
+    # If not set and create is true, a name is generated using the fullname
+    name:
+
+socketDir: /var/lib/kubelet/plugins/cephfs.csi.ceph.com
```
this is cephfs socket directory not rbd

Madhu-1 commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -0,0 +1,215 @@
+socketFile: csi.sock
+registrationDir: /var/lib/kubelet/plugins_registry
+pluginDir: /var/lib/kubelet/plugins
+driverName: cephfs.csi.ceph.com
```
same for driver name

wilmardo replied:
😳 copy paste errors, sorry about that. Fixed all references now :)

Madhu-1 replied:
still, I see the issue
```
[🎩︎]mrajanna@localhost ceph-csi-rbd $]grep cephfs -ri *
values.yaml:socketDir: /var/lib/kubelet/plugins/cephfs.csi.ceph.com
values.yaml:driverName: cephfs.csi.ceph.com
[🎩︎]mrajanna@localhost ceph-csi-rbd $]pwd
/home/mrajanna/workspace/src/github.com/ceph/ceph-csi/charts/ceph-csi-rbd
[🎩︎]mrajanna@localhost ceph-csi-rbd $]

```

```
[🎩︎]mrajanna@localhost ceph-csi-cephfs $]grep rbd -ri *
Chart.yaml:  - https://github.com/ceph/ceph-csi/tree/master/charts/ceph-csi-rbd
README.md:kubectl delete namespace ceph-csi-rbd
```

wilmardo replied:
Sure you pullled the last commit?

```console
~/Development/Prive/ceph-csi grep cephfs -ri charts/ceph-csi-rbd/*
~/Development/Prive/ceph-csi grep rbd -ri charts/ceph-csi-cephfs/*
```

Madhu-1 replied:
nope its not present now, Thanks

Madhu-1 commented:
@wilmardo can you do final testing, for me one for the pod is in Crash state due to port issue

```
[🎩︎]mrajanna@localhost ceph-csi-cephfs $]kubectl get po -nceph-csi-cephfs
NAME                                           READY   STATUS             RESTARTS   AGE
ceph-csi-cephfs-nodeplugin-q7dq8               1/3     CrashLoopBackOff   8          2m8s
ceph-csi-cephfs-provisioner-7b4cc86854-4fhng   4/4     Running            1          2m8s
ceph-csi-cephfs-provisioner-7b4cc86854-bnlkf   4/4     Running            1          2m8s
ceph-csi-cephfs-provisioner-7b4cc86854-hr2ct   4/4     Running            1          2m8s
[🎩︎]mrajanna@localhost ceph-csi-cephfs $]
[🎩︎]mrajanna@localhost ceph-csi-cephfs $]kubectl get po -nceph-csi-rbd
NAME                                        READY   STATUS    RESTARTS   AGE
ceph-csi-rbd-nodeplugin-kk8vq               3/3     Running   0          2m30s
ceph-csi-rbd-provisioner-694f4b545d-29xmw   5/5     Running   2          2m30s
ceph-csi-rbd-provisioner-694f4b545d-k77mf   5/5     Running   2          2m30s
ceph-csi-rbd-provisioner-694f4b545d-ks5jf   5/5     Running   2          2m30s
```

wilmardo commented:
> can you do final testing, for me one for the pod is in Crash state due to port issue

What is the kubectl describe? I can't test the cephfs pods since my Ceph cluster is at 14.1.x

Madhu-1 commented:
> > can you do final testing, for me one for the pod is in Crash state due to port issue
> 
> What is the kubectl describe? I can't test the cephfs pods since my Ceph cluster is at 14.1.x

just deploy cephfs and create PV, even PV creation fails for it no issue

wilmardo commented:
> just deploy cephfs and create PV, even PV creation fails for it no issue

Reason is that the rbd and cephfs poth expose metrics on the same ports, switched cephfs to 9080 and 9090. The liveness cannot bind on the port since it is already in use:
```
kubectl log -n default csi-cephfsplugin-4kxpt liveness-prometheus
I0911 12:25:44.296053       1 cephcsi.go:103] Driver version: canary and Git version: f4b38228ae7f2fae6313683008c106e4dddbc8da
I0911 12:25:44.296284       1 cephcsi.go:158] Starting driver type: liveness with name: liveness.csi.ceph.com
I0911 12:25:44.296307       1 liveness.go:84] Liveness Running
F0911 12:25:44.296590       1 httpserver.go:25] listen tcp 192.168.2.155:8081: bind: address already in use
```

All the deployment files in the current master seem to have this problem, I deployed `deploy/cephfs/kubernetes/v1.14` alongside rbd with metrics and exact the problem as above.
This is now fixed in the helm chart, if #604 gets merged soon after the problem is done :)

Madhu-1 commented:
current ports
* rbd
```
 provisioner and nodeplugin
metrics 8090
liveness 8080

- name: http-metrics
      port: 8080
      protocol: TCP
      targetPort: 8080
    - name: grpc-metrics
      port: 8090
      protocol: TCP
      targetPort: 8090
```

* cephfs
```
metrics 8091
liveness 8081

service
- name: http-metrics
      port: 8080
      protocol: TCP
      targetPort: 8081
    - name: grpc-metrics
      port: 8090
      protocol: TCP
      targetPort: 8091
```

Madhu-1 commented:
@wilmardo any update on this?

wilmardo commented:
Had a busy day yesterday, will start looking into this again right now.

wilmardo commented:
Last commit fixes the last issues. @Madhu-1 thanks for the overview!
Tested it locally on kind and all pods came up (integration isn't tested yet).
```
NAMESPACE     NAME                                                            READY   STATUS    RESTARTS   AGE    IP           NODE                 NOMINATED NODE   READINESS GATES
default       flippant-buffalo-ceph-csi-rbd-nodeplugin-8khjh                  3/3     Running   0          53s    172.17.0.3   kind-worker2         <none>           <none>
default       flippant-buffalo-ceph-csi-rbd-nodeplugin-hgb6h                  3/3     Running   0          53s    172.17.0.2   kind-worker          <none>           <none>
default       flippant-buffalo-ceph-csi-rbd-provisioner-d6956c966-jlzv4       5/5     Running   0          53s    10.244.2.5   kind-worker2         <none>           <none>
default       flippant-buffalo-ceph-csi-rbd-provisioner-d6956c966-l68kf       5/5     Running   0          53s    10.244.1.5   kind-worker          <none>           <none>
default       flippant-buffalo-ceph-csi-rbd-provisioner-d6956c966-z7zfk       5/5     Running   0          53s    10.244.2.4   kind-worker2         <none>           <none>
default       washing-cardinal-ceph-csi-cephfs-nodeplugin-9727w               3/3     Running   0          38s    172.17.0.3   kind-worker2         <none>           <none>
default       washing-cardinal-ceph-csi-cephfs-nodeplugin-vlqt9               3/3     Running   0          38s    172.17.0.2   kind-worker          <none>           <none>
default       washing-cardinal-ceph-csi-cephfs-provisioner-5779877746-6llhk   4/4     Running   0          38s    10.244.2.6   kind-worker2         <none>           <none>
default       washing-cardinal-ceph-csi-cephfs-provisioner-5779877746-85btb   4/4     Running   0          38s    10.244.1.7   kind-worker          <none>           <none>
default       washing-cardinal-ceph-csi-cephfs-provisioner-5779877746-tm676   4/4     Running   0          38s    10.244.1.6   kind-worker          <none>           <none>
```

Madhu-1 commented:
@wilmardo Thanks for your patience, I will do the integration test and will  do a final review

Madhu-1 commented:
@wilmardo I deployed this on one node kubernetes cluster
am able to create PVC with both cephfs and rbd
am able to mount PVC with rbd 

things not working

* liveness container is not working for both rbd and cephfs (issue with socket file)
* driver registrar is not able to register cephfs( because of this PVC mounting failing)


I would suggest you check the socketfile implementation in kubernetes folder, in  Kubernetes  we create 2 sockets
* provisioner socket https://github.com/ceph/ceph-csi/blob/master/deploy/cephfs/kubernetes/v1.14%2B/csi-cephfsplugin-provisioner.yaml#L101

* nodeplugin socket https://github.com/ceph/ceph-csi/blob/master/deploy/cephfs/kubernetes/v1.14%2B/csi-cephfsplugin.yaml#L79

and also the volumes in kubernetes templates and helm chats are not matching please check the volumes in both cephfs and rbd (refer kubernetes templates)

code changes I made for cephfs to working  [cephfs.txt](https://github.com/ceph/ceph-csi/files/3620698/cephfs.txt)

once the deployment is done please check logs of all container for verification

Madhu-1 commented:
@wilmardo https://github.com/ceph/ceph-csi/pull/620 same changes have to be addressed here also

Madhu-1 commented:
@wilmardo any update on this one?

wilmardo commented:
@Madhu-1  No update just yet, hopefully I have time today but work is busy at the moment :) Certainly will make time this week, this shouldn't linger any longer.

Madhu-1 commented:
@wilmardo Thanks no issue

wilmardo commented:
Just tested cephfs and rbd as statefulset and deployment, seems to work!

## Cephfs

To generate the templates:
`helm template --name ceph-csi-cephfs charts/ceph-csi-cephfs --kube-version 1.14 > cephfs.yml`

`helm template --name ceph-csi-cephfs charts/ceph-csi-cephfs --kube-version 1.13 > cephfs-1.13.yml`

## RBD
To generate the templates:
`helm template --name ceph-csi-rbd charts/ceph-csi-rbd --kube-version 1.14 > rbd.yml` 

`helm template --name ceph-csi-rbd charts/ceph-csi-rbd --kube-version 1.13 > rbd-1.13.yml` 

# How to review

The I diffed the current deployment files in the master with hunks of the templates in https://www.diffchecker.com/diff (or any difftool for that matter). I my opinion it looks good, I thinks this currently is the best way to review this PR since Github doesn't show the changes that well.

Madhu-1 commented:
will review it tomorrow, @wilmardo Thanks for updating PR

Madhu-1 commented:
Few points
* for statefulset replica should be one
* attacher address is wrong (cephfs) https://github.com/ceph/ceph-csi/blob/d96d3a01261f125eb4a794c9867f4dd7d582ad5b/charts/ceph-csi-cephfs/templates/provisioner-statefulset.yaml#L58 need to be changed to `{{ .Values.provisionerSocketFile }}`

@wilmardo need these changes

Madhu-1 commented:
once changes are done, please squash the commits Thanks

wilmardo commented:
Done and done!

I think you meant `"unix:///csi/{{ .Values.provisionerSocketFile }}"`?

Madhu-1 commented:
> Done and done!
> 
> I think you meant `"unix:///csi/{{ .Values.provisionerSocketFile }}"`?

`{{ .Values.provisionerSocketFile }}` instead of `{{ .Values.socketFile }}`

Madhu-1 commented:
hardcode replica count  to 1 for statefulset, we canot have more  replica  for statefulset

wilmardo commented:
> hardcode replica count to 1 for statefulset, we canot have more replica for statefulset

Will do!

> {{ .Values.provisionerSocketFile }} instead of {{ .Values.socketFile }}

https://github.com/ceph/ceph-csi/blob/8805237bd7d4caf78935e0fcc793af1361fd15fd/charts/ceph-csi-cephfs/templates/provisioner-statefulset.yaml#L58

That should be it right? That is the ADDRESS value for all the other envs

Madhu-1 commented:
> > hardcode replica count to 1 for statefulset, we canot have more replica for statefulset
> 
> Will do!
> 
> > {{ .Values.provisionerSocketFile }} instead of {{ .Values.socketFile }}
> 
> https://github.com/ceph/ceph-csi/blob/8805237bd7d4caf78935e0fcc793af1361fd15fd/charts/ceph-csi-cephfs/templates/provisioner-statefulset.yaml#L58
> 
> That should be it right? That is the ADDRESS value for all the other envs

both works, but need to test it, I have tested  ` {{ .Values.provisionerSocketFile }}` works fine

Madhu-1 commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -0,0 +1,224 @@
+# The filename of the plugin socket
+pluginSocketFile: csi.sock
+# The directory where nodeplugin socker is created
+registrationDir: /var/lib/kubelet/plugins_registry
```
socker to socket

Madhu-1 commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -0,0 +1,231 @@
+provisionerSocketFile: csi-provisioner.sock
+# The filename of the plugin socket
+pluginSocketFile: csi.sock
+# The directory where nodeplugin socker is created
```
socker to socket

Madhu-1 commented:
@wilmardo let me know once the  fix is done, i will do final testing before approval

wilmardo commented:
@Madhu-1 Done!

Madhu-1 commented:
doing testing, once things look fine, will approve it

Madhu-1 commented:
we need to backport it to release branch once this is merged

humblec commented:
Awesome !! Thanks .. LGTM.. Approved!
